lib/std/io_uring.zig

```diff
@@ -0,0 +1,841 @@
+    pub fn init_params(entries: u32, p: *io_uring_params) !IO_Uring {
+        assert(entries >= 1 and entries <= 4096 and std.math.isPowerOfTwo(entries));
+        assert(p.*.sq_entries == 0);
+        assert(p.*.cq_entries == 0);
+        assert(p.*.features == 0);
+        assert(p.*.wq_fd == 0);
+        assert(p.*.resv[0] == 0);
+        assert(p.*.resv[1] == 0);
+        assert(p.*.resv[2] == 0);
+        if (!supported) return error.IO_UringKernelNotSupported;
+
+        const res = linux.io_uring_setup(entries, p);
+        try check_errno(res);
+        const fd = @intCast(i32, res);
+        assert(fd >= 0);
+        errdefer os.close(fd);
+
+        // Kernel versions 5.4 and up use only one mmap() for the submission and completion queues.
+        // This is not an optional feature for us... if the kernel does it, we have to do it.
+        // The thinking on this by the kernel developers was that both the submission and the
+        // completion queue rings have sizes just over a power of two, but the submission queue ring
+        // is significantly smaller with u32 slots. By bundling both in a single mmap, the kernel
+        // gets the submission queue ring for free.
+        // See https://patchwork.kernel.org/patch/11115257 for the kernel patch.
+        // We do not support the double mmap() done before 5.4, because we want to keep the
+        // init/deinit mmap paths simple and because io_uring has had many bug fixes even since 5.4.
+        if ((p.*.features & linux.IORING_FEAT_SINGLE_MMAP) == 0) {
+            return error.IO_UringKernelNotSupported;
+        }
+
+        // Check that the kernel has actually set params and that "impossible is nothing".
+        assert(p.*.sq_entries != 0);
+        assert(p.*.cq_entries != 0);
+        assert(p.*.cq_entries >= p.*.sq_entries);
+
+        // From here on, we only need to read from params, so pass `p` by value for convenience.
+        // The completion queue shares the mmap with the submission queue, so pass `sq` there too.
+        var sq = try SubmissionQueue.init(fd, p.*);
+        errdefer sq.deinit();
+        var cq = try CompletionQueue.init(fd, p.*, sq);
+        errdefer cq.deinit();
+
+        // Check that our starting state is as we expect.
+        assert(sq.head.* == 0);
+        assert(sq.tail.* == 0);
+        assert(sq.mask.* == p.*.sq_entries - 1);
+        // Allow flags.* to be non-zero, since the kernel may set IORING_SQ_NEED_WAKEUP at any time.
+        assert(sq.dropped.* == 0);
+        assert(sq.array.len == p.*.sq_entries);
+        assert(sq.sqes.len == p.*.sq_entries);
+        assert(sq.sqe_head == 0);
+        assert(sq.sqe_tail == 0);
+
+        assert(cq.head.* == 0);
+        assert(cq.tail.* == 0);
+        assert(cq.mask.* == p.*.cq_entries - 1);
+        assert(cq.overflow.* == 0);
+        assert(cq.cqes.len == p.*.cq_entries);
+
+        // Alles in Ordnung!
```

Comment:
?

Reply:
"Everything in order"... a play on words. This was part of the early dev for a code walkthrough and slipped through here. Removed.